### ISSUE TITLE
Dimension inconsistencies for setting/getting values?

### ISSUE THREAD

**Issue by tamasgal** [JuliaIO/HDF5.jl#551]
Maybe it's just me, but I find it a bit weird that a `dataset` with a given shape returns a an array with the same shape but one entry at `[1,1,1,...]`. However setting it accepts a scalar.

To be more specific, I am working with large histograms (10^12 bins) and use HDF5 to create an empty dataset and update the bin counts by accessing them directly, since for obvious reasons it's not feasible to load the file into the RAM.

I assumed this would work, but it doesn't since I get back an array:

```julia
using HDF5

h5open("big_file.h5", "cw") do file
    if !has(file, "data")
        d_create(file, "data", datatype(Int32), dataspace(2, 3, 4))
    end
    dset = file["data"]
    
    dset[1,2,3] += 1
end
```

This is the error:

```
MethodError: no method matching +(::Array{Int32,3}, ::Int64)
Closest candidates are:
  +(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:502
  +(!Matched::Complex{Bool}, ::Real) at complex.jl:292
  +(!Matched::Missing, ::Number) at missing.jl:97
  ...
```

Then I figured I should directly target it, which at first seemed to work but it didn't update the file:

```julia
h5open("big_file.h5", "cw") do file
    if !has(file, "data")
        d_create(file, "data", datatype(Int32), dataspace(2, 3, 4))
    end
    dset = file["data"]
    
    dset[1,2,3][1,1,1] += 1
end

# returns 1, due to the last expression

# however, the file is not modified:

h5open("big_file.h5", "r") do file
    println(file["data"][:,:,:])
end

Int32[0 0 0; 0 0 0]

Int32[0 0 0; 0 0 0]

Int32[0 0 0; 0 0 0]

Int32[0 0 0; 0 0 0]
```

Finally the working solution, which looks really ugly for higher dimensions:

```julia
h5open("big_file.h5", "cw") do file
    if !has(file, "data")
        d_create(file, "data", datatype(Int32), dataspace(2, 3, 4))
    end
    dset = file["data"]
    
    v = dset[1,2,3][1,1,1]
    dset[1,2,3] = v + 1
end

h5open("big_file.h5", "r") do file
    println(file["data"][:,:,:])
end

Int32[0 0 0; 0 0 0]

Int32[0 0 0; 0 0 0]

Int32[0 1 0; 0 0 0]

Int32[0 0 0; 0 0 0]
```

I think the first approach would be the most natural one. Or what do you think?

**Comment by rafaqz** [comment]
I missed this - it's basically the same issue I raise in #576. Standard behaviour in `getindex` in julia is that a returned array will drop the dimension of integer indices, and passing all integer indices will return a scalar (or a zero dimension array for `view`)

**Comment by tamasgal** [comment]
Yes exactly. I don't know how to solve this, as it will definitely break a lot of code ;)

**Comment by rafaqz** [comment]
Time for a major version bump :) 

It shouldn't be much work really, either `dropdims` on the returned array from `_getindex` or pass in a `ret` array with dimensions of length 1 removed (not sure if that will work with the underlying copy mechanism?)
https://github.com/JuliaIO/HDF5.jl/blob/master/src/HDF5.jl#L1774

**Comment by musm** [comment]
Thanks, I agree we should fix this and make a new   potentially breaking major release

**Comment by musm** [comment]
We are planning on making a breaking release, should we try to work towards this for that?

**Comment by kleinhenz** [comment]
closed by #603. inplace modification syntax seems to work now as well.
